internal/pkg/upload/upload.go

```diff
@@ -0,0 +1,339 @@
+				if !chunkT.Stop() {
+					<-chunkT.C
+				}
+				chunkT.Reset(chunkProgressTimeout)
+			case <-complete: // upload operation complete, clean up
+				if !chunkT.Stop() {
+					<-chunkT.C
+				}
+				if !total.Stop() {
+					<-total.C
+				}
+				u.finalize(id)
+				return
+			}
+		}
+	}()
+	info := Info{
+		ID:        id,
+		DocID:     docID,
+		ChunkSize: MaxChunkSize,
+		Source:    source,
+		Total:     size,
+		Hasher:    hasher,
+		HashSum:   hashsum,
+	}
+	cnt := info.Total / info.ChunkSize
+	if info.Total%info.ChunkSize > 0 {
+		cnt += 1
+	}
+	info.Count = int(cnt)
+	u.current[id] = upload{
+		opToken:       token,
+		chunkThrottle: throttle.NewThrottle(u.parallelChunkLimit),
+		complete:      complete,
+		chunkRecv:     chunkRecv,
+		Info:          info,
+	}
+	return info, nil
+}
+
+func (u *Uploader) Chunk(uplID string, chunkID int) (ChunkInfo, error) {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+	upl, valid := u.current[uplID]
+	if !valid {
+		return ChunkInfo{}, ErrInvalidUploadID
+	}
+	if chunkID < 0 || chunkID >= upl.Info.Count {
+		return ChunkInfo{}, errors.New("invalid chunk number")
+	}
+
+	token := upl.chunkThrottle.Acquire(strconv.Itoa(chunkID), time.Hour)
+	if token == nil {
+		return ChunkInfo{}, ErrMaxConcurrentUploads
+	}
+	upl.chunkRecv <- struct{}{}
+	if !upl.begun {
+		upl.begun = true
+	}
+	u.current[uplID] = upl
+
+	return ChunkInfo{
+		ID:            chunkID,
+		FirstReceived: upl.begun,
+		Final:         chunkID == upl.Info.Count-1,
+		Upload:        upl.Info,
+		Token:         token,
+	}, nil
+}
+
+func (u *Uploader) Complete(id string, bulker bulk.Bulk) (Info, error) {
+	info, valid := u.current[id]
+	if !valid {
+		return Info{}, ErrInvalidUploadID
+	}
+
+	ok, err := u.allChunksPresent(info.Info, bulker)
+	if err != nil {
+		return Info{}, err
+	}
+	if !ok {
+		return Info{}, ErrMissingChunks
+	}
+
+	ok, err = u.verifyChunkData(info.Info, bulker)
+	if err != nil {
+		return Info{}, err
+	}
+	if !ok {
+		return Info{}, errors.New("file contents did not pass validation")
+	}
+
+	// @todo: verify chunks
+	// verify hashes, etc
+	u.current[id].complete <- struct{}{}
+	return info.Info, nil
+}
+
+func (u *Uploader) cleanupOperation(uplID string) {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+	if upload, ok := u.current[uplID]; ok {
+		if upload.opToken != nil {
+			upload.opToken.Release()
+		}
+	}
+	delete(u.current, uplID)
+}
+
+func (u *Uploader) cancel(uplID string) error {
+	u.cleanupOperation(uplID)
+
+	// @todo: delete any uploaded chunks from ES
+	// leave header doc and mark failed?
+	return nil
+}
+
+func (u *Uploader) finalize(uplID string) error {
+	u.cleanupOperation(uplID)
+	// @todo: write Status:READY here?
+	return nil
+}
+
+func (u *Uploader) allChunksPresent(info Info, bulker bulk.Bulk) (bool, error) {
+	hits, err := ListChunkIDs(context.TODO(), bulker, info.Source, info.DocID)
+	if err != nil {
+		log.Warn().Err(err).Msg("error listing chunks")
+		return false, err
+	}
+	if len(hits) != info.Count {
```

Comment:
think you might hit this place if you have more than 10 chunks. need to use info.Count as "size" in the query to get the max expected number of hits